quinn-proto/src/endpoint.rs

```diff
@@ -542,44 +529,124 @@ impl Endpoint {
+        let tls = server_config
+            .crypto
+            .clone()
+            .start_session(incoming.version, &params);
         let transport_config = server_config.transport.clone();
         let mut conn = self.add_connection(
             ch,
-            version,
-            dst_cid,
+            incoming.version,
+            incoming.dst_cid,
             loc_cid,
-            src_cid,
-            addresses,
+            incoming.src_cid,
+            incoming.addresses,
             now,
             tls,
             Some(server_config),
             transport_config,
+            incoming.retry_src_cid.is_some(),
         );
-        if dst_cid.len() != 0 {
-            self.index.insert_initial(dst_cid, ch);
+        if incoming.dst_cid.len() != 0 {
+            self.index.insert_initial(incoming.dst_cid, ch);
         }
-        match conn.handle_first_packet(now, addresses.remote, ecn, packet_number, packet, rest) {
+        match conn.handle_first_packet(
+            now,
+            incoming.addresses.remote,
+            incoming.ecn,
+            incoming.packet_number,
+            incoming.packet,
+            incoming.rest,
+        ) {
             Ok(()) => {
-                trace!(id = ch.0, icid = %dst_cid, "connection incoming");
-                Some(DatagramEvent::NewConnection(ch, conn))
+                trace!(id = ch.0, icid = %incoming.dst_cid, "new connection");
+                Ok((ch, conn))
             }
             Err(e) => {
                 debug!("handshake failed: {}", e);
                 self.handle_event(ch, EndpointEvent(EndpointEventInner::Drained));
                 match e {
-                    ConnectionError::TransportError(e) => Some(DatagramEvent::Response(
-                        self.initial_close(version, addresses, crypto, &src_cid, e, buf),
-                    )),
-                    _ => None,
+                    ConnectionError::TransportError(e) => Err(Some(self.initial_close(
+                        incoming.version,
+                        incoming.addresses,
+                        &incoming.crypto,
+                        &incoming.src_cid,
+                        e,
+                        buf,
+                    ))),
+                    _ => Err(None),
                 }
             }
         }
     }
 
+    /// Reject this incoming connection attempt.
+    pub fn reject(&mut self, incoming: IncomingConnection, buf: &mut BytesMut) -> Transmit {
+        self.initial_close(
+            incoming.version,
+            incoming.addresses,
+            &incoming.crypto,
+            &incoming.src_cid,
+            TransportError::CONNECTION_REFUSED(""),
+            buf,
+        )
+    }
+
+    /// Respond with a retry packet, requiring the client to retry with address validation.
+    ///
+    /// Panics if `may_retry` is false.
+    pub fn retry(&mut self, incoming: IncomingConnection, buf: &mut BytesMut) -> Transmit {
```

Comment:
Rather than requiring the user to call `may_retry` and panicing if they get it wrong, could we make `may_retry` private and return an error here if it's `false`? That way it's impossible for users to forget to check.

Reply:
Done (both in quinn-proto and in quinn).